fix(core): resolve error for multiple component instances that use fallback content

crisbeto opened the pull request:
Currently fallback content for `ng-content` gets declared and rendered out in one go. This breaks down if multiple instances of the same component are used where one doesn't render the fallback content while the other one does, because the `TNode` for the content has to be created during the first creation pass.

These changes resolve the issue by always _declaring_ the template, but only rendering it if the slot is empty.

Fixes #55466.

AndrewKushnir commented on packages/core/src/render3/instructions/projection.ts:
```diff
@@ -161,13 +171,10 @@ export function ɵɵprojection(
-  const fallbackTNode =
-      declareTemplate(lView, tView, fallbackIndex, templateFn, decls, vars, null, attrs);
+function insertFallbackContent(lView: LView, tView: TView, fallbackIndex: number) {
+  const fallbackTNode = tView.data[HEADER_OFFSET + fallbackIndex] as TNode;
   const fallbackLContainer = lView[HEADER_OFFSET + fallbackIndex];
```
```suggestion
  const fallbackTemplateIndex = HEADER_OFFSET + fallbackIndex;
  const fallbackTNode = tView.data[fallbackTemplateIndex] as TNode;
  const fallbackLContainer = lView[fallbackTemplateIndex];
```

crisbeto replied:
We use the `adjustedIndex` terminology in other places so I'll go with that.

crisbeto committed:
fix(core): resolve error for multiple component instances that use fallback content

Currently fallback content for `ng-content` gets declared and rendered out in one go. This breaks down if multiple instances of the same component are used where one doesn't render the fallback content while the other one does, because the `TNode` for the content has to be created during the first creation pass.

These changes resolve the issue by always _declaring_ the template, but only rendering it if the slot is empty.

Fixes #55466.

crisbeto commented:
Caretaker note: there's one unrelated failing target in the presubmit. Also these changes passed a presubmit earlier today.

AndrewKushnir commented:
This PR was merged into the repository by commit 97eea8d50ea1be55850c43df58b27a1aa5557326.
